lib/srv/desktop/tdp/conn.go

```diff
@@ -185,138 +200,190 @@ func IsFatalErr(err error) bool {
+				// In that case, we'll try again via the loop.
+				h.msgs, err = i(m)
 				if err != nil {
-					return trace.Wrap(err)
-				}
-			}
-			if msg != nil {
-				err := c.SendToClient(msg)
-				if err != nil {
-					return trace.Wrap(err)
+					return nil, err
 				}
+				h.msgs = removeNilMessages(h.msgs)
+				h.next = 0
+			} else {
+				// No interceptor to run
+				return m, err
 			}
 		}
-	})
 
-	// Run a goroutine to read TDP messages coming from the client
-	// and pass them on to the Windows agent.
-	errs.Go(func() error {
-		defer closeAll()
+		// Return the next cached message
+		nextMsg := h.msgs[h.next : h.next+1]
+		h.next++
+		return nextMsg[0], nil
+	}
+}
 
-		for {
-			msg, err := c.client.ReadMessage()
-			if err := c.handleError(err); err != nil {
-				return err
-			}
+// NewReadWriteInterceptor creates a new 'ReadWriteInterceptor' that intercepts messages on 'src'.
+// The provided interceptor callbacks may be nil.
+func NewReadWriteInterceptor(src MessageReadWriteCloser, readInterceptor, writeInterceptor Interceptor) *ReadWriteInterceptor {
+	return &ReadWriteInterceptor{
+		src:              src,
+		writeInterceptor: writeInterceptor,
+		readAdapter:      readInterceptorAdapter(src, readInterceptor),
+	}
+}
 
-			if err := c.SendToServer(msg); err != nil {
-				return trace.Wrap(err)
+// WriteMessage passes the message to the write interceptor (if provided)
+// for omition or modification before writing the message to the underlying
+// writer.
+func (i *ReadWriteInterceptor) WriteMessage(m Message) error {
+	var err error
+	var out []Message
+	if i.writeInterceptor != nil {
+		out, err = i.writeInterceptor(m)
+		if err == nil {
+			for _, msg := range removeNilMessages(out) {
+				err = i.src.WriteMessage(msg)
+				if err != nil {
+					break
+				}
 			}
 		}
-	})
-
-	// Wait for all goroutines to finish
-	if err := errs.Wait(); err != nil && !utils.IsOKNetworkError(err) {
-		return trace.Wrap(err)
+		return err
 	}
+	// No interceptor found
+	return i.src.WriteMessage(m)
+}
 
-	return nil
+// ReadMessage reads from the underlying reader and passes them to the
+// read interceptor (if provided) for omition or modification before
+// returning the next message.
+func (i *ReadWriteInterceptor) ReadMessage() (Message, error) {
+	return i.readAdapter()
 }
 
-func (c *ConnProxy) handleError(err error) error {
-	if err == nil {
-		return nil
-	}
-	if utils.IsOKNetworkError(err) {
-		return trace.Wrap(err)
+// Close calls close on the underlying 'MessageReadWriteCloser'
+func (i *ReadWriteInterceptor) Close() error {
+	return i.src.Close()
+}
+
+// messageCopy behaves similarly to io.Copy except it deals with Message types.
+// It reads messages from 'src' and writes them to 'dst' until an error is received.
+// It does *not* forward an EOF received from the reader, but returns nil in the happy path.
+func messageCopy(dst MessageWriter, src MessageReader) error {
+	var err error
```

Comment:
We should never see `io.EOF` with valid message (it would be decode message end nil err + io.EOF on next read) so this method can be written as: 
```go 
for {
	msg, err := src.ReadMessage()
	if errors.Is(err, io.EOF) {
		return nil
	} else if err != nil {
		return err
	}
	if err := dst.WriteMessage(msg); err != nil {
		return err
	}
}
```

Reply:
For me it is 100%, you don't have to track where `err` is set and you have regular error handling like you're used to: call - handle error - another call - handle error